pkg/kwokctl/runtime/binary/cluster.go

```diff
@@ -0,0 +1,530 @@
+	if err != nil {
+		return err
+	}
+
+	if conf.PrometheusPort != 0 {
+		prometheusPortStr := strconv.Itoa(int(conf.PrometheusPort))
+
+		prometheusData, err := BuildPrometheus(BuildPrometheusConfig{
+			ProjectName:               conf.Name,
+			SecretPort:                conf.SecretPort,
+			AdminCrtPath:              adminCertPath,
+			AdminKeyPath:              adminKeyPath,
+			PrometheusPort:            int(conf.PrometheusPort),
+			EtcdPort:                  etcdClientPort,
+			KubeApiserverPort:         apiserverPort,
+			KubeControllerManagerPort: kubeControllerManagerPort,
+			KubeSchedulerPort:         kubeSchedulerPort,
+			KwokPort:                  kwokControllerPort,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to generate prometheus yaml: %s", err)
+		}
+		prometheusConfigPath := utils.PathJoin(conf.Workdir, runtime.Prometheus)
+		err = os.WriteFile(prometheusConfigPath, []byte(prometheusData), 0644)
+		if err != nil {
+			return fmt.Errorf("failed to write prometheus yaml: %s", err)
+		}
+
+		prometheusPath := utils.PathJoin(bin, "prometheus")
+		prometheusArgs := []string{
+			"--config.file",
+			prometheusConfigPath,
+			"--web.listen-address",
+			serveAddress + ":" + prometheusPortStr,
+		}
+		err = utils.ForkExec(ctx, conf.Workdir, prometheusPath, prometheusArgs...)
+		if err != nil {
+			return err
+		}
+	}
+
+	// set the context in default kubeconfig
+	c.Kubectl(ctx, utils.IOStreams{}, "config", "set", "clusters."+conf.Name+".server", scheme+"://"+localAddress+":"+apiserverPortStr)
+	c.Kubectl(ctx, utils.IOStreams{}, "config", "set", "contexts."+conf.Name+".cluster", conf.Name)
+	if conf.SecretPort {
+		c.Kubectl(ctx, utils.IOStreams{}, "config", "set", "clusters."+conf.Name+".insecure-skip-tls-verify", "true")
+		c.Kubectl(ctx, utils.IOStreams{}, "config", "set", "contexts."+conf.Name+".user", conf.Name)
+		c.Kubectl(ctx, utils.IOStreams{}, "config", "set", "users."+conf.Name+".client-certificate", adminCertPath)
+		c.Kubectl(ctx, utils.IOStreams{}, "config", "set", "users."+conf.Name+".client-key", adminKeyPath)
+	}
+	return nil
+}
+
+func (c *Cluster) Down(ctx context.Context) error {
+	conf, err := c.Config()
+	if err != nil {
+		return err
+	}
+
+	c.Kubectl(ctx, utils.IOStreams{}, "config", "unset", "clusters."+conf.Name)
+	c.Kubectl(ctx, utils.IOStreams{}, "config", "unset", "users."+conf.Name)
+	c.Kubectl(ctx, utils.IOStreams{}, "config", "unset", "contexts."+conf.Name)
+
+	bin := utils.PathJoin(conf.Workdir, "bin")
+	kubeApiserverPath := utils.PathJoin(bin, "kube-apiserver"+vars.BinSuffix)
+	kubeControllerManagerPath := utils.PathJoin(bin, "kube-controller-manager"+vars.BinSuffix)
+	kubeSchedulerPath := utils.PathJoin(bin, "kube-scheduler"+vars.BinSuffix)
+	kwokControllerPath := utils.PathJoin(bin, "kwok-controller"+vars.BinSuffix)
+	etcdPath := utils.PathJoin(bin, "etcd")
+	prometheusPath := utils.PathJoin(bin, "prometheus")
+
+	err = utils.ForkExecKill(ctx, conf.Workdir, kwokControllerPath)
```

Comment:
for later: maybe we can think about doing the `ForkExecKill` of each component in parallel rather than serial (same for bring up)